core-rust/state-manager/src/state_manager.rs

```diff
@@ -712,123 +711,132 @@ where
 
         // Whilst we could validate intent hash duplicates here, these are checked by validators on prepare already,
         // and the check will move into the engine at some point and we'll get it for free then...
-        let prepared_transactions: Vec<_> = commit_request
+        let prepare_results = commit_request
             .transactions
-            .into_iter()
-            .map(|raw| -> (RawLedgerTransaction, PreparedLedgerTransaction) {
-                let prepared = self.ledger_transaction_validator.prepare_from_raw(&raw)
-                    .unwrap_or_else(|error| {
-                        panic!("Committed transaction cannot be prepared - likely byzantine quorum: {error:?}");
-                    });
-                (raw, prepared)
+            .iter()
+            .map(|raw_transaction| {
+                self.ledger_transaction_validator
+                    .prepare_from_raw(raw_transaction)
             })
-            .collect();
+            .collect::<Result<Vec<_>, _>>();
+
+        let prepared_transactions = match prepare_results {
+            Ok(prepared_transactions) => prepared_transactions,
+            Err(error) => {
+                warn!("cannot parse transaction to be committed {error:?}",);
+                return Err(InvalidCommitRequestError::TransactionParsingFailed);
+            }
+        };
 
         let mut write_store = self.store.write();
         let mut series_executor = self.start_series_execution(write_store.deref());
 
         if commit_request_start_state_version != series_executor.latest_state_version() {
             panic!(
-                "Mismatched state versions - the commit request claims {} but the database thinks we're at {}",
-                commit_request_start_state_version, series_executor.latest_state_version()
+                "ledger state version differs from the proof ({} vs ({} - {} == {}))",
+                series_executor.latest_state_version(),
+                commit_state_version,
+                commit_transactions_len,
+                commit_request_start_state_version
             );
         }
 
+        // TODO(optimization-only): we could reach to the `execution_cache` to find the transaction
+        // root chain there (to avoid this "repeated" calculation here).
+        let resultant_transaction_root = Self::calculate_transaction_root(
```

Comment:
_(note from author)_

Here is some unnecessary computation (for @malexandru-rdx's attention). There are a couple of ways to mitigate it (none of them perfect; actually the most tempting of them would create a "malicious actor can trigger an OOME" risk).
I went for simplicity, and I indicated my best mitigation idea in that TODO.